dnf5/commands/download/download.cpp

```diff
@@ -112,47 +116,52 @@ void DownloadCommand::configure() {
-        package_query.resolve_pkg_spec(option->get_value(), {}, true);
-        package_query.filter_priority();
-        package_query.filter_latest_evr();
-        result_query |= package_query;
-    }
-
-    if (resolve_option->get_value()) {
-        auto goal = std::make_unique<libdnf::Goal>(ctx.base);
-        libdnf::GoalJobSettings settings;
-        for (auto pkg : result_query) {
-            goal->add_rpm_install(pkg, settings);
-        }
-        auto transaction = goal->resolve();
-        auto transaction_problems = transaction.get_problems();
-        if (transaction_problems != libdnf::GoalProblem::NO_PROBLEM) {
-            if (transaction_problems != libdnf::GoalProblem::NOT_FOUND) {
-                throw GoalResolveError(transaction);
-            }
-        }
-        download_pkgs.reserve(transaction.get_transaction_packages().size());
-        for (auto & tspkg : transaction.get_transaction_packages()) {
-            if (transaction_item_action_is_inbound(tspkg.get_action()) &&
-                tspkg.get_package().get_repo()->get_type() != libdnf::repo::Repo::Type::COMMANDLINE) {
-                download_pkgs.push_back(tspkg.get_package());
+        pkg_query.resolve_pkg_spec(option->get_value(), {}, true);
+        pkg_query.filter_priority();
+        pkg_query.filter_latest_evr();
+
+        for (const auto & pkg : pkg_query) {
+            if (resolve_option->get_value()) {
+                auto goal = std::make_unique<libdnf::Goal>(ctx.base);
+                goal->add_rpm_install(pkg, {});
+
+                auto transaction = goal->resolve();
+                auto transaction_problems = transaction.get_problems();
+                if (transaction_problems != libdnf::GoalProblem::NO_PROBLEM) {
+                    if (transaction_problems != libdnf::GoalProblem::NOT_FOUND) {
+                        throw GoalResolveError(transaction);
+                    }
+                }
+                for (auto & tspkg : transaction.get_transaction_packages()) {
+                    if (transaction_item_action_is_inbound(tspkg.get_action()) &&
+                        tspkg.get_package().get_repo()->get_type() != libdnf::repo::Repo::Type::COMMANDLINE) {
+                        download_pkgs.insert(tspkg.get_package());
+                    }
+                }
+            } else {
+                download_pkgs.insert(std::move(pkg));
             }
         }
-    } else {
-        download_pkgs.reserve(result_query.size());
-        for (auto package : result_query) {
-            download_pkgs.push_back(std::move(package));
-        }
     }
 
     if (!download_pkgs.empty()) {
-        download_packages(download_pkgs, destdir_option->get_value().c_str());
+        libdnf::repo::PackageDownloader downloader(ctx.base);
+
+        // for download command, we don't want to mark the packages for removal
+        downloader.force_caching(true);
+
+        for (auto & pkg : download_pkgs) {
```

Comment:
It will sound strange, but we cannot download all packages. We have to download only one package per NEVRA or per file name. The reason is that we download files into one particular directories. It means that RPMs will override each other and because we download them in parallel, the download will fail due to incorrect checksum.

DNF4 implementation:
```
        pkg_dict = {}
        for pkg in pkgs:
            pkg_dict.setdefault(str(pkg), []).append(pkg)

        to_download = []
        cmdline = []
        for pkg_list in pkg_dict.values():
            pkgs_cmdline = [pkg for pkg in pkg_list
                            if pkg.repoid == hawkey.CMDLINE_REPO_NAME]
            if pkgs_cmdline:
                cmdline.append(pkgs_cmdline[0])
                continue
            pkg_list.sort(key=lambda x: (x.repo.priority, x.repo.cost))
            to_download.append(pkg_list[0])
```